accepted/future-releases/class-modifiers/feature-specification.md

```diff
@@ -530,62 +578,264 @@ mixinModifier     ::= 'sealed' | 'base' | 'interface' | 'final'
+
+    base class D implements S {} // Error, cannot use "implements".
+    ```
+
+### Mixin restrictions
+
+There are a few changes around mixins to support `mixin class` and disallow
+using normal `class` declarations as mixins while dealing with language
+versioning and backwards compatibility.
+
+Currently, a class may only be used as a mixin if it has a default constructor.
+This prevents the class from defining a `const` constructor or any factory
+constructors. We loosen this somewhat. Define a *trivial generative constructor*
+to be a generative constructor that:
+
+*   Is not a redirecting constructor,
+
+*   declares no parameters,
+
+*   has no initializer list (no `: ...` part),
+
+*   has no body (only `;`), and
+
+*   is not `external`. *An `external` constructor is considered to have an
+    externally provided initializer list and/or body.*
+
+A trivial constructor may be named or unnamed, and `const` or non-`const`. A
+*non-trivial generative constructor* is a generative constructor which is not a
+trivial generative constructor. Examples:
+
+```dart
+class C {
+  // Trivial generative constructors:
+  C();
+  const C();
+
+  // Non-trivial generative constructors:
+  C(int x);
+  C(this.x);
+  C() { ... }
+  C(): super();
+
+  // Not generative constructors, so neither trivial generative nor non-trivial
+  // generative:
+  factory C.f = C;
+  factory C.f2() { ... }
+}
+```
+
+It's a compile-time error if:
+
+*   A `mixin class` declaration has a superclass other than `Object`. *The
+    declaration is limited to an `extends Object` clause or no `extends` clause,
+    and no `with` clauses. The class grammar prohibits `on` clauses.*
+
+*   A `mixin class` declaration declares any non-trivial generative constructor.
+    *It may declare no constructors, in which case it gets a default
+    constructor, or it can declare factory constructors and/or trivial
+    generative constructors.*
+
+These rules ensure that when you mark a `class` with `mixin` that it *can* be
+used as one.
+
+A class not marked `mixin` can still be used as a mixin when the class's
+declaration is in a pre-feature library. Post-feature libraries can also mix in
+their own classes that aren't marked `mixin` as long as the class supports it.
+Specifically:
+
+It's a compile-time error to for a declaration in library `L` to mix in a
+non-`mixin` class declaration `D` from library `K` if any of:
+
+*   The superclass of `D` is not `Object`,
+
+*   `K` is a pre-feature library, and `D` declares any constructors, or
+
+    *For pre-feature libraries, we cannot tell if the intent of `class` was
+    "just a class" or "both a class and a mixin". For compatibility, we assume
+    the latter, even if the class is being used as a mixin in a post-feature
+    library where it does happen to be possible to distinguish those two
+    intents.*
+
+*   `K` is a post-feature library, and any of:
+
+    *   `K` is not the same library as `L`,
```

Comment:
Can we change the spec to talk about how only mixin classes and mixins can be mixed in regardless of library since this mornings meeting? And that its an error for non-mixin classes to be mixed in?
Not sure if I got that right...

Reply:
Should be done now.

Basically, it's a compile-time error to mix in a non-`mixin` `class` unless:
    * that class is declared in a pre-feature library,
    * it has `Object` as superclass, and
    * it declares no constructors.

(The same rules that apply to mixing in a `class` declaration in 2.19, where all libraries are pre-feature.)

There are now no other exceptions, even inside the same library as the class.